src/azul/plugins/metadata/hca/transform.py

```diff
@@ -1408,7 +1408,7 @@ def _matrices_for(category: SubmitterCategory) -> JSONs:
             return [
                 self._matrices(file, submitter)
                 for file, submitter in matrices
-                if submitter.category == category
+                if submitter is not None and submitter.category == category
```

Comment:
In that case, 

```diff
Index: src/azul/plugins/metadata/hca/transform.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/metadata/hca/transform.py b/src/azul/plugins/metadata/hca/transform.py
--- a/src/azul/plugins/metadata/hca/transform.py	(revision eec7c106f85ed338ed6d08c0b4b3ac6e2cfabf9c)
+++ b/src/azul/plugins/metadata/hca/transform.py	(date 1626890583484)
@@ -1399,16 +1399,16 @@
         project = self._get_project(self.api_bundle)
 
         matrices = [
-            (file, Submitter.for_file(file))
+            (file, Submitter.category_for_file(file))
             for file in visitor.files.values()
             if file.is_matrix
         ]
 
         def _matrices_for(category: SubmitterCategory) -> JSONs:
             return [
-                self._matrices(file, submitter)
-                for file, submitter in matrices
-                if submitter is not None and submitter.category == category
+                self._matrices(file, category)
+                for file, category_ in matrices
+                if category_ == category
             ]
 
         contents = dict(self._samples(samples.values()),
```

Reply:
Also, remind me again, which matrix files do not have a submitter?

Reply:
The canned bundle `fce68057-b0f0-5d11-b9a7-30e8fa3259a8` has analysis files matrices that don't have a `submitter_id` or `file_source` field, however the changes being made in #3096 ( https://github.com/DataBiosphere/azul/pull/3134/files ) will recognize these files as intermediate matrices and apply the 'DCP/2 Analysis' submitter value to them.

I've queried the `..._file` tables in the snapshots used by catalogs dcp1, dcp6, and dcp7 and wasn't able to locate a file that had "matrix" in its `content_description` and that was missing a `submitter_id`, other than analysis files.

Question: do you still want this patch applied, or should this ticket be blocked on #3096? (Note: #3096 itself is blocked on #3192 which is currently under peer-review.)

Reply:
> I've queried the ..._file tables in the snapshots used by catalogs dcp1, dcp6, and dcp7 and wasn't able to locate a file that had "matrix" in its content_description and that was missing a submitter_id, other than analysis files.

When I say "have a submitter" I really meant a `submitter_id` or a `file_source`. Those `analysis_file` rows, are they intermediate analysis matrices or organically described CGMs? For the former, we hard-code a submitter, for the latter we derive the submitter from `file_source`. If there are any other `analysis_file` rows that have neither `submitter_id` nor `file_source` I would like to know.

I prefer to have submitters, not categories as the second tuple element in the `matrices` list as long as the submitters are guaranteed to never be None. I see two options: 1) applying the patch and adding a FIXME for 3096 reminding us to switch back to submitters or 2) blocking as you describe and holding this PR warm. Your call but I think the FIXME is the prudent thing to do.

Reply:
> If there are any other analysis_file rows that have neither submitter_id nor file_source I would like to know.

There are 1,035 in dcp7

```
# Matrix with neither 'submitter_id' nor 'file_source': 1,035
SELECT analysis_file_id FROM `broad-datarepo-terra-prod-hca2.hca_prod_20201120_dcp2___20210707_dcp7.analysis_file` WHERE REGEXP_CONTAINS(JSON_EXTRACT(content, "$.file_core.content_description"), r'(?i)matrix') AND content NOT LIKE '%submitter_id%' AND content NOT LIKE '%file_source%'

# Matrix with 'file_source': 11
SELECT analysis_file_id FROM `broad-datarepo-terra-prod-hca2.hca_prod_20201120_dcp2___20210707_dcp7.analysis_file` WHERE REGEXP_CONTAINS(JSON_EXTRACT(content, "$.file_core.content_description"), r'(?i)matrix') AND content LIKE '%file_source%'

# Matrix with 'submitter_id': 82
SELECT analysis_file_id FROM `broad-datarepo-terra-prod-hca2.hca_prod_20201120_dcp2___20210707_dcp7.analysis_file` WHERE REGEXP_CONTAINS(JSON_EXTRACT(content, "$.file_core.content_description"), r'(?i)matrix') AND content LIKE '%submitter_id%'
```

Reply:
Can you determine if those 1035 rows all represent intermediate matrices generated by DCP/2 analysis? The most reliable way is to join against the process and protocols that produced them, but there may be another distinguishing marker in the `content` column of these rows. Could you post an example of the `content` in one of those 1035 rows?

Reply:
1,025 of the rows have the same `content_description` as this example:
```
{
    "describedBy":"https://schema.humancellatlas.org/type/file/6.2.0/analysis_file",
    "file_core":{
        "content_description":[
            {
                "ontology":"data:3917",
                "ontology_label":"Count Matrix",
                "text":"DCP/2-generated matrix"
            }
        ],
        "file_name":"27e2b162-443a-4c09-9061-d01ad378f40d.loom",
        "format":"loom"
    },
    "provenance":{
        "document_id":"35d54cbf-fe81-5a8f-a82f-319d9d4c5e48",
        "submission_date":"2021-05-24T12:00:00.000000Z"
    },
    "schema_type":"file"
}
```

The other 10 have
```
{
    "text":"Gene expression matrix",
    "ontology":"data:3112",
    "ontology_label":"Gene expression matrix"
}
```

Reply:
So it looks like we have 

- 82 project-level DCP/2 matrices, 
- 11 organically described CGMs and 
- 1025 intermediate DCP/2 matrices. 

The remaining 10 `analysis_file` rows look like organic CGMs that erroneously don't have a `file_source` even though they appear to be from GEO. The changes in #3096 would treat them as if they were intermediate DCP/2 matrices. The code on `prod` as of today will index them into the `/index/files` endpoint but exclude them from any `matrices` or `contributorMatrices` trees on the `/indexprojects endpoint`. Could you please determine the project ID for those 10 `analysis_file` rows? It is possible that these rows are "garbage", not referenced by any `links` rows.

Reply:
```
# 10 anaysis files with 'data:3112' and no 'submitter_id' nor 'file_source'
SELECT analysis_file_id FROM `broad-datarepo-terra-prod-hca2.hca_prod_20201120_dcp2___20210707_dcp7.analysis_file`
WHERE REGEXP_CONTAINS(JSON_EXTRACT(content, "$.file_core.content_description"), r'(?i)matrix') 
AND content NOT LIKE '%submitter_id%' AND content NOT LIKE '%file_source%' and content like '%data:3112%'

analysis_file_id
da86d37d-0dca-4fc7-81d8-12882a6c487d
7a878935-fc9d-423c-843e-7578882500e1
c3cec7a4-73e8-485b-8c39-d5828976aa41
e389da75-5725-4ea6-8a56-fdf20acf126a
768eb998-ca69-4315-82f0-21616617e10e
4434e9b2-799b-4288-bf05-d14957a5b401
773efd75-510b-4614-bd24-2e4715db678a
d09eed31-9694-42fd-ba92-a464d0ddd136
d4e2ff66-847c-450f-b912-37c5668f7fcf
9ff93ab7-a63a-4f07-88ea-8746dd0759e3

# 10 subgraphs containing the 10 analysis files
SELECT links_id, project_id FROM `broad-datarepo-terra-prod-hca2.hca_prod_20201120_dcp2___20210707_dcp7.links`
where REGEXP_CONTAINS(content, r'da86d37d-0dca-4fc7-81d8-12882a6c487d|e389da75-5725-4ea6-8a56-fdf20acf126a|768eb998-ca69-4315-82f0-21616617e10e|4434e9b2-799b-4288-bf05-d14957a5b401|773efd75-510b-4614-bd24-2e4715db678a|d09eed31-9694-42fd-ba92-a464d0ddd136|d4e2ff66-847c-450f-b912-37c5668f7fcf|9ff93ab7-a63a-4f07-88ea-8746dd0759e3|7a878935-fc9d-423c-843e-7578882500e1|c3cec7a4-73e8-485b-8c39-d5828976aa41')
ORDER BY project_id, links_id

links_id                                project_id
10ebb2c5-3f73-436f-bf43-aece8e7e25a6    51f02950-ee25-4f4b-8d07-59aa99bb3498
c8089427-dc20-4ee6-9876-1c29d47b0ba4    51f02950-ee25-4f4b-8d07-59aa99bb3498
4029c7a5-ecf3-48f2-8dd3-45c4aaca08b8    d3ac7c1b-5302-4804-b611-dad9f89c049d
4514694e-a1ef-4c29-b4f5-246f6baf6845    d3ac7c1b-5302-4804-b611-dad9f89c049d
615335eb-0dbd-44cb-831a-a454d4e0627e    d3ac7c1b-5302-4804-b611-dad9f89c049d
b34f5177-1cb1-4feb-a351-170f6cd4e60c    d3ac7c1b-5302-4804-b611-dad9f89c049d
e5c13eda-e583-44a1-915f-85da7bc3114c    d3ac7c1b-5302-4804-b611-dad9f89c049d
e8892faa-5e4a-4396-8f02-235f141b28aa    d3ac7c1b-5302-4804-b611-dad9f89c049d
f17f1464-f938-4c8d-bfe9-5c08f86b2c28    d3ac7c1b-5302-4804-b611-dad9f89c049d
702be28e-3d95-4812-a8d9-24ffdae7211a    df88f39f-01a8-4b5b-92f4-3177d6c0f242

# In each case, the analysis file is an output from a 'process' type process with cell suspension inputs

analysis_file   7a878935-fc9d-423c-843e-7578882500e1
links_id        10ebb2c5-3f73-436f-bf43-aece8e7e25a6
project_id      51f02950-ee25-4f4b-8d07-59aa99bb3498

analysis_file   c3cec7a4-73e8-485b-8c39-d5828976aa41
links_id        c8089427-dc20-4ee6-9876-1c29d47b0ba4
project_id      51f02950-ee25-4f4b-8d07-59aa99bb3498

analysis_file   d09eed31-9694-42fd-ba92-a464d0ddd136
links_id        4029c7a5-ecf3-48f2-8dd3-45c4aaca08b8
project_id      d3ac7c1b-5302-4804-b611-dad9f89c049d

analysis_file   4434e9b2-799b-4288-bf05-d14957a5b401
links_id        4514694e-a1ef-4c29-b4f5-246f6baf6845
project_id      d3ac7c1b-5302-4804-b611-dad9f89c049d

analysis_file   e389da75-5725-4ea6-8a56-fdf20acf126a
links_id        615335eb-0dbd-44cb-831a-a454d4e0627e
project_id      d3ac7c1b-5302-4804-b611-dad9f89c049d

analysis_file   9ff93ab7-a63a-4f07-88ea-8746dd0759e3
links_id        b34f5177-1cb1-4feb-a351-170f6cd4e60c
project_id      d3ac7c1b-5302-4804-b611-dad9f89c049d

analysis_file   773efd75-510b-4614-bd24-2e4715db678a
links_id        e5c13eda-e583-44a1-915f-85da7bc3114c
project_id      d3ac7c1b-5302-4804-b611-dad9f89c049d

analysis_file   768eb998-ca69-4315-82f0-21616617e10e
links_id        e8892faa-5e4a-4396-8f02-235f141b28aa
project_id      d3ac7c1b-5302-4804-b611-dad9f89c049d

analysis_file   d4e2ff66-847c-450f-b912-37c5668f7fcf
links_id        f17f1464-f938-4c8d-bfe9-5c08f86b2c28
project_id      d3ac7c1b-5302-4804-b611-dad9f89c049d

analysis_file   da86d37d-0dca-4fc7-81d8-12882a6c487d
links_id        702be28e-3d95-4812-a8d9-24ffdae7211a
project_id      df88f39f-01a8-4b5b-92f4-3177d6c0f242
```

Reply:
Please raise stub issue about this or let me know why it would be correct and expected to have `analysis_file` matrices without either `file_source` or `submitter_id`.

Reply:
Created stub https://github.com/DataBiosphere/azul/issues/3297